### ISSUE TITLE
Authentication issue during terraform refresh with token from module output

### ISSUE THREAD

**Issue by Jasper-Ben** [gitlabhq/terraform-provider-gitlab#1071]
### GitLab Provider version

3.14.0

### GitLab version

14.9.2-ee

### Terraform version

v1.1.8

### Relevant Terraform Configuration

```hcl
terraform {
  required_version = ">= 0.15"
  required_providers {
    gitlab = {
      source  = "gitlabhq/gitlab"
      version = ">= 3.14.0"
    }
  }
}

provider "gitlab" {
  base_url         = "https://${module.k8s_gitlab.gitlab_url}/api/v4/"
  token            = module.k8s_gitlab.gitlab_api_token
  # early_auth_check disabled, as the availability of the token depends on a module.
  early_auth_check = false
}

data "gitlab_user" "root" {
  username = "root"
}

resource "gitlab_personal_access_token" "runner_registry_access" {
  user_id = data.gitlab_user.root.user_id
  name    = "runner_registry_access"
  scopes  = ["write_registry"]
}
```

### Relevant log output

```shell
[0m[1mgitlab_personal_access_token.runner_registry_access: Refreshing state... [id=1:90][0m
2022-05-03T17:09:27.126+0200 [INFO]  provider.terraform-provider-gitlab_v3.14.0: 2022/05/03 17:09:27 [DEBUG] read gitlab PersonalAccessToken 90, user ID 1: timestamp=2022-05-03T17:09:27.126+0200
2022-05-03T17:09:27.126+0200 [INFO]  provider.terraform-provider-gitlab_v3.14.0: 2022/05/03 17:09:27 [DEBUG] GitLab API Request Details:
---[ REQUEST ]---------------------------------------
GET /api/v4/personal_access_tokens?page=1&per_page=100&user_id=1 HTTP/1.1
Host: <gitlab_url>
User-Agent: Terraform/1.1.8 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-gitlab/3.14.0
Accept: application/json
Authorization: Bearer
Accept-Encoding: gzip


-----------------------------------------------------: timestamp=2022-05-03T17:09:27.126+0200
2022-05-03T17:09:27.169+0200 [INFO]  provider.terraform-provider-gitlab_v3.14.0: 2022/05/03 17:09:27 [DEBUG] GitLab API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 401 Unauthorized
Content-Length: 30
Cache-Control: no-cache
Content-Type: application/json
Date: Tue, 03 May 2022 15:09:27 GMT
Vary: Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Request-Id: 01G257G57585HWDREQ7XM90ES7
X-Runtime: 0.009575

{
 "message": "401 Unauthorized"
}
-----------------------------------------------------: timestamp=2022-05-03T17:09:27.169+0200
2022-05-03T17:09:27.169+0200 [ERROR] vertex "gitlab_personal_access_token.runner_registry_access" error: GET https://<gitlab_url>/api/v4/personal_access_tokens: 401 {message: 401 Unauthorized}
2022-05-03T17:09:27.170+0200 [ERROR] vertex "gitlab_personal_access_token.runner_registry_access" error: GET https://<gitlab_url>/api/v4/personal_access_tokens: 401 {message: 401 Unauthorized}
2022-05-03T17:09:27.170+0200 [ERROR] vertex "gitlab_personal_access_token.runner_registry_access (expand)" error: GET https://<gitlab_url>/api/v4/personal_access_tokens: 401 {message: 401 Unauthorized}

```

### Description

:wave: 

### Expected behavior:
gitlab provider should be able to authenticate successfully every time when interacting with the GitLab API.

### Actual behaviour:
While the initial authentication works as expected and the gitlab provider successfully creates the registry access token, running subsequent `terraform refresh`'s will cause a 401 Unauthorized error message.

**Comment by Jasper-Ben** [comment]
I don't understand how this issue does not affect everyone. I mean, this is just core functionality? Or am I being stupid?

**Comment by Jasper-Ben** [comment]
So I have tried hard-coding the token and now it works. Bummer :disappointed: Is this a limitation in terraform itself?

**Comment by timofurrer** [comment]
Is the execution order correct? Meaning is `gitlab_user` really only executed after the `module.k8s_gitlab.gitlab_api_token` is available? 

Can you try setting an explicit dependency?

```hcl
terraform {
  required_version = ">= 0.15"
  required_providers {
    gitlab = {
      source  = "gitlabhq/gitlab"
      version = ">= 3.14.0"
    }
  }
}

provider "gitlab" {
  base_url         = "https://${module.k8s_gitlab.gitlab_url}/api/v4/"
  token            = module.k8s_gitlab.gitlab_api_token
  # early_auth_check disabled, as the availability of the token depends on a module.
  early_auth_check = false
}

data "gitlab_user" "root" {
  username = "root"

  depends_on = [module.k8s_gitlab]
}

resource "gitlab_personal_access_token" "runner_registry_access" {
  user_id = data.gitlab_user.root.user_id
  name    = "runner_registry_access"
  scopes  = ["write_registry"]
}
```

**Comment by timofurrer** [comment]
what bugs me is that your posted error is from the `gitlab_personal_access_token` resource ... therefore, did the `gitlab_user` run fine?

Actually, what token is `gitlab_api_token`?

Remember it has to be an administrator token, see [docs](https://docs.gitlab.com/ee/api/users.html#:~:text=Use%20this%20API%20to%20create%20a%20new%20personal%20access%20token.%20Token%20values%20are%20returned%20once%20so%2C%20make%20sure%20you%20save%20it%20as%20you%20can%E2%80%99t%20access%20it%20again.%20This%20API%20can%20only%20be%20used%20by%20GitLab%20administrators.)

> Use this API to create a new personal access token. Token values are returned once so, make sure you save it as you can’t access it again. This API can only be used by GitLab administrators.

**Comment by Jasper-Ben** [comment]
> Is the execution order correct? Meaning is gitlab_user really only executed after the module.k8s_gitlab.gitlab_api_token is available?

Should be, due to the implicit dependency... and as I said it works during the initial creation. So I'm currently thinking that this isn't an issue during the initial apply because the resource doesn't exist yet and thus does not get refreshed. However, during subsequent applies terraform tries to refresh the resource and it seems that the value of a module output might not yet be available during the refresh phase.

> Can you try setting an explicit dependency?

Will try. Setting an explicit `depends_on` on the `gitlab_user` data source might actually help...

> Also see: https://github.com/hashicorp/terraform/issues/2430

Yeah I remember (actually commented on that issue on a separate occasion a while ago). Was already about to dive into reading that issue again when you replied.

**Comment by Jasper-Ben** [comment]
> what bugs me is that your posted error is from the `gitlab_personal_access_token` resource ... therefore, did the `gitlab_user` run fine?
> 
> Actually, what token is `gitlab_api_token`?
> 
> Remember it has to be an administrator token, see [docs](https://docs.gitlab.com/ee/api/users.html#:~:text=Use%20this%20API%20to%20create%20a%20new%20personal%20access%20token.%20Token%20values%20are%20returned%20once%20so%2C%20make%20sure%20you%20save%20it%20as%20you%20can%E2%80%99t%20access%20it%20again.%20This%20API%20can%20only%20be%20used%20by%20GitLab%20administrators.)
> 
> > Use this API to create a new personal access token. Token values are returned once so, make sure you save it as you can’t access it again. This API can only be used by GitLab administrators.

The token definitely is correct. As I said, the resources are created correctly during the first apply.

**Comment by timofurrer** [comment]
> The token is definitely correct. As I said, the resources are created correctly during the first apply. 

oh, I've misread that part, my bad.

Where is the `gitlab_api_token` coming from? Is it available during `refresh` / `plan` ?

Does `terraform apply -refresh-only` work instead of the `refresh` ?

**Comment by Jasper-Ben** [comment]
> oh, I've misread that part, my bad.

no worries :slightly_smiling_face: 

> Where is the gitlab_api_token coming from? Is it available during refresh / plan ?

 It's the output of a module that already has been deployed. So the token should already be available in the tfstate.

> Does terraform apply -refresh-only work instead of the refresh ?

It does. :o That's interesting... any idea why this might be the case?

**Comment by timofurrer** [comment]
It's just that the `terraform refresh` command is [deprectated](https://www.terraform.io/cli/commands/refresh#:~:text=Warning%3A%20This%20command%20is%20deprecated%2C%20because%20its%20default%20behavior%20is%20unsafe%20if%20you%20have%20misconfigured%20credentials%20for%20any%20of%20your%20providers.%20See%20below%20for%20more%20information%20and%20recommended%20alternatives.).

I can tell you exactly what the problem in this case is though, but it mentions some misconfiguration with provider configurations - not sure if that is applicable here 🤷 

Are you okay with closing this and just using `terraform apply -refresh-only` instead?

**Comment by Jasper-Ben** [comment]
the problem is, that `terraform apply` does an implicit `terraform refresh` as well. So this also breaks during apply.

**Comment by timofurrer** [comment]
Hmmm, can you provide a fully runnable example which reproduces the issue?

Otherwise, I'll need some time to reproduce this ..

**Comment by Jasper-Ben** [comment]
So far this is what I found out:

works:
- `terraform apply -refresh-only`
- `terraform apply -refresh=false`

does not work:
- `terraform apply`
- `terraform refresh`

So my best guess would be that `terraform apply` still uses the deprecated `terraform refresh` under the hood.

**Comment by Jasper-Ben** [comment]
> Hmmm, can you provide a fully runnable example which reproduces the issue?

Sure, let me try.

**Comment by Jasper-Ben** [comment]
Hm. So I am actually struggling to get a minimal "breaking" example. This is what I've tried so far and it works as intended:
https://github.com/Jasper-Ben/example_gitlab_token_terraform

So no idea (yet) why it breaks in our specific usecase. Unfortunately it isn't easy to reproduce the exact setup without further complex dependencies (AWS, Kubernetes). Will have take another look at this tomorrow and post some more details...

**Comment by Jasper-Ben** [comment]
So the good news is that the issue does not seem to come from using module output values as token alone. Which gives me hope that this might be solvable.

**Comment by timofurrer** [comment]
When you run a plan is the `gitlab_api_token` "only known after apply" ? e.g. a computed field based on something only known after apply?

**Comment by timofurrer** [comment]
> So the good news is that the issue does not seem to come from using module output values as token alone. Which gives me hope that this might be solvable.

Yes, and I doubt that it's actually a bug in this provider, but rather a limitation in terraform itself - which, if solvable, needs some workaround ...

**Comment by Jasper-Ben** [comment]
Managed to solve this :slightly_smiling_face: 

Before, the token came from a module output which got its value from a data source. Terraform didn't seem to like that, I guess since the data output is fetched dynamically during each refresh.

I've changed my setup so that the value actually comes from a terraform resource definition, now things seem to work.


More boring details on the exact setup:

I had a Kubernetes job which would programmatically generate an API token, add it to GitLab and store it into a Kubernetes secret. I then used a kubernetes secret data-source for returning the token as an output:
```
output "api_token" {
  value = data.kubernetes_secret ...
  sensitive = true
}
```
Now I've moved the token generation part into terraform instead, which so far seems to work.
```
resource "random_password" "api_token" {
  length           = 20
  special          = true
}
output "api_token" {
  value = random_password.api_token.result
  sensitive = true
}
```

Thanks for your help @timofurrer :heart:
Closing.
